src/dune_engine/process.ml

```diff
@@ -372,17 +372,21 @@ module Exit_status = struct
 
   (* In this module, we don't need the "Error: " prefix given that it is already
      included in the error message from the command. *)
-  let fail ~dir paragraphs =
+  let fail ~dir ~run_cancelled paragraphs =
     let dir =
       match dir with
       | None -> Path.of_string (Sys.getcwd ())
       | Some dir -> dir
     in
-    raise
-      (User_error.E
-         (User_message.make paragraphs, Some (With_directory_annot.make dir)))
+    let exn =
+      User_error.E
+        (User_message.make paragraphs, Some (With_directory_annot.make dir))
+    in
+    match run_cancelled with
+    | false -> raise exn
+    | true -> raise (Memo.Non_reproducible exn)
```

Comment:
Should we actually expect `exn` to be useful in this case? Maybe we should instead raise something like `Memo.Non_reproducible (Failure "Build cancelled")`, instead, so that when the error propagates to the root, without the `Non_reproducible` marker, it's less surprising?

Reply:
Indeed, done.